### ISSUE TITLE
iCloud sync stops when zone gets deleted

### ISSUE THREAD

**Issue by yoasha** [insidegui/WWDC#509]
I'm running WWDC app version 6.0.4 on macOS Mojave on two devices: iMac and MacBook Pro and app data is out of sync (i.e. favotites  and session progress). Sync used to work properly in the past but currently it isn't. I'm not sure if issue is related to macOS Mojave...

I tried to restart both computers, disable and enable back iCloud sync in preferences, but nothing helped. I also wanted to clear iCloud data but strangely I couldn't find WWDC iCloud data in macOS System Preferences > iCloud > Manage.

I'm not sure if this is a general sync bug or just a local issue specific to my environment. Any suggestions of how to diagnose would be appreciated.

**Comment by insidegui** [comment]
You can help us diagnose the issue by running the app with Console.app running, filtering by `WWDC` and sending us a txt file with the logs.

**Comment by yoasha** [comment]
Thanks for your reply. Attached two log files: one log file when WWDC app is loaded, and another log file when I context click a session and selected "Add to Favorites".
[WWDC_add_favorite.txt](https://github.com/insidegui/WWDC/files/2473216/WWDC_add_favorite.txt)
[WWDC_load.txt](https://github.com/insidegui/WWDC/files/2473217/WWDC_load.txt)

**Comment by insidegui** [comment]
Hmmm, there's an error message `"Zone was purged by user"`, so it looks like you deleted the iCloud data for the WWDC app at some point, which wiped out the custom zone we use for syncing. We'll need to handle that error in our sync code.

**Comment by yoasha** [comment]
I probably deleted it by mistake, though I can't remember doing it. Is there any workaround I can manually apply now, or this issue must be handled by WWDC app's code?

**Comment by insidegui** [comment]
You could try deleting the preferences by running `defaults delete io.wwdc.app` in Terminal and restarting the app.

**Comment by yoasha** [comment]
Two issues:
1. After deleting preferences as you instructed, there are warnings and errors when loading app. See attached file.

Specifically:
WARNINGS (YELLOW)
error	10:12:41.508386 +0300	WWDC	Unable to load Info.plist exceptions (eGPUOverrides)
default	10:12:41.627270 +0300	WWDC	background session setup reply received: session <private> with identifier <private>
error	10:12:42.017486 +0300	WWDC	        AEGetDescData of non-desc type 'reco' not a good idea

ERRORS (RED)
fault	10:12:42.748139 +0300	WWDC	Favorite <private> is not associated to a session. That's illegal!
fault	10:12:42.748202 +0300	WWDC	Favorite <private> is not associated to a session. That's illegal!
fault	10:12:42.748364 +0300	WWDC	Favorite <private> is not associated to a session. That's illegal!
fault	10:12:42.748541 +0300	WWDC	Favorite <private> is not associated to a session. That's illegal!
fault	10:12:42.748718 +0300	WWDC	Favorite <private> is not associated to a session. That's illegal!


2. Deleting preferences did create the iCloud data again when I enabled the "iCloud beta" on UI. However, I can now understand why I previously deleted the WWDC iCloud data, as it has blank icon and named as "app". As I'm a developer testing many demo apps, I frequently deleting unused apps from iCloud to keep my list clean and maintainable, and I mistakenly considered WWDC app as a demo app. I guess that a related issue should be added to project.
[WWDC warnings and errors on load after deleting defaults.txt](https://github.com/insidegui/WWDC/files/2476158/WWDC.warnings.and.errors.on.load.after.deleting.defaults.txt)
<img width="780" alt="screen shot 2018-10-14 at 9 49 47" src="https://user-images.githubusercontent.com/5601548/46913604-2f946380-cf99-11e8-909e-bfb4d23b08fa.png">

**Comment by insidegui** [comment]
Just reiterating that this is a bug where the app won't re-create the CloudKit zone if it gets deleted, we need to handle this case in the sync code.

**Comment by yoasha** [comment]
I understand. Is there anyway I can manually fix or workaround this issue rather than wait for a fix via the code?

**Comment by insidegui** [comment]
Make sure the app is not running, then run the following commands in Terminal:

```
defaults delete io.wwdc.app createdPrivateSubscription
defaults delete io.wwdc.app privateChangeToken
defaults delete io.wwdc.app createdCustomZone
```

That should make the app re-create the missing zone and re-upload everything once it's launched again.

**Comment by yoasha** [comment]
Thank you. It worked perfectly!

**Comment by insidegui** [comment]
This has been addressed in version 7.1
